### PR TITLE
Fewer queries when caching hearing request types

### PR DESCRIPTION
**Pull request by hschallhorn**
Resolves some of the runtime issues for the update_cached_appeals_attributes_job

### Description
Do not build an appeal for each vacols case we cache. This produced 4 calls to vacols for every single appeal we cache and makes the runtime of the job quite high.

```ruby
...
[2020-10-21 16:45:07 -0400]   LegacyAppeal Load (0.4ms)  SELECT  "legacy_appeals".* FROM "legacy_appeals" WHERE "legacy_appeals"."vacols_id" = $1 LIMIT $2  [["vacols_id", "6210367160"], ["LIMIT", 1]]
[2020-10-21 16:45:07 -0400]   VACOLS::Correspondent Load (3.0ms)  SELECT  "CORRES".* FROM "CORRES" WHERE "CORRES"."STAFKEY" = :a1 FETCH FIRST :a2 ROWS ONLY  [["stafkey", "1972342395"], ["LIMIT", 1]]
[2020-10-21 16:45:07 -0400]   VACOLS::Folder Load (3.1ms)  SELECT  "FOLDER".* FROM "FOLDER" WHERE "FOLDER"."TICKNUM" = :a1 FETCH FIRST :a2 ROWS ONLY  [["ticknum", "6210367160"], ["LIMIT", 1]]
[2020-10-21 16:45:07 -0400]   VACOLS::CaseIssue Load (4.1ms)  SELECT "ISSUES".* FROM "ISSUES" WHERE "ISSUES"."ISSKEY" = :a1  [["isskey", "6210367160"]]
...
# Four of these calls for each appeal we're caching
```

```ruby
vacols_ids = LegacyAppeal.pluck(:vacols_id)
# only 31 appeals

def former_travel?(legacy_appeal)
  # the current request type is travel
  if legacy_appeal.current_hearing_request_type == :travel_board
    return false
  end

  # otherwise check if og request type was travel
  legacy_appeal.original_hearing_request_type == :travel_board
end

def old_call(vacols_ids)
  VACOLS::Case.where(bfkey: vacols_ids).pluck(:bfkey, :bfac, :bfcurloc).map do |bfkey, bfac, bfcurloc|
    [
      bfkey,
      {
        location: bfcurloc,
        status: VACOLS::Case::TYPES[bfac]
      }
    ]
  end.to_h
end

def new_call(vacols_ids)
  VACOLS::Case.where(bfkey: vacols_ids).map do |vacols_case|
    legacy_appeal = AppealRepository.build_appeal(vacols_case) # build non-persisting legacy appeal object

    [
      vacols_case.bfkey,
      {
        location: vacols_case.bfcurloc,
        status: VACOLS::Case::TYPES[vacols_case.bfac],
        hearing_request_type: legacy_appeal.current_hearing_request_type(readable: true),
        former_travel: former_travel?(legacy_appeal)
      }
    ]
  end.to_h
end

def new_new_call(vacols_ids)
  VACOLS::Case.where(bfkey: vacols_ids).map do |vacols_case|
    original_request = original_hearing_request_type_for_vacols_case(vacols_case)
    changed_request = changed_hearing_request_type_for_all_vacols_ids[vacols_case.bfkey]
    current_request = HearingDay::REQUEST_TYPES.key(changed_request)&.to_sym || original_request

    [
      vacols_case.bfkey,
      {
        location: vacols_case.bfcurloc,
        status: VACOLS::Case::TYPES[vacols_case.bfac],
        hearing_request_type: LegacyAppeal::READABLE_HEARING_REQUEST_TYPES[current_request],
        former_travel: original_request == :travel_board && current_request != :travel_board
      }
    ]
  end.to_h
end

def original_hearing_request_type_for_vacols_case(vacols_case)
  request_type = VACOLS::Case::HEARING_REQUEST_TYPES[vacols_case.bfhr]

  (request_type == :travel_board && vacols_case.bfdocind == "V") ? :video : request_type
end

def changed_hearing_request_type_for_all_vacols_ids
  @changed_hearing_request_type_for_all_vacols_ids ||= LegacyAppeal.pluck(:vacols_id, :changed_request_type).to_h
end

Benchmark.measure { old_call(vacols_ids) }
=> #<Benchmark::Tms:0x00007ff5c6cd9100
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=0.3174909995868802,
 @stime=0.017438999999999538,
 @total=0.07847399999999816,
 @utime=0.06103499999999862>

Benchmark.measure { new_call(vacols_ids) }
=> #<Benchmark::Tms:0x00007ff5c16471e8
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=2.2098000003024936,
 @stime=0.14571100000000037,
 @total=1.3108650000000015,
 @utime=1.1651540000000011>

Benchmark.measure { new_new_call(vacols_ids) }
=> #<Benchmark::Tms:0x00007fec37238ba8
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=0.4909099997021258,
 @stime=0.029617000000000004,
 @total=0.19788100000000064,
 @utime=0.16826400000000064>
```

### Acceptance Criteria
- [ ] Cached appeal attributes are still correct
- [ ] Runtime of update cached appeal attributes job is shorter

### Testing Plan
1. on master in dev
```ruby
Benchmark.measure { UpdateCachedAppealsAttributesJob.new.perform_now }
=> #<Benchmark::Tms:0x00007fd1cdaed138
 @cstime=0.045598,
 @cutime=0.03593299999999999,
 @label="",
 @real=4.578323000110686,
 @stime=0.26611799999999985,
 @total=3.2173869999999996,
 @utime=2.869738>
```
1. on this branch in dev, exit console and come back in
```ruby
Benchmark.measure { UpdateCachedAppealsAttributesJob.new.perform_now }
=> #<Benchmark::Tms:0x00007ff557c029a8
 @cstime=0.033809,
 @cutime=0.032311999999999994,
 @label="",
 @real=3.9726599999703467,
 @stime=0.2307260000000002,
 @total=2.7981360000000004,
 @utime=2.501289>
```

#### If you're feeling fun
1. In prod
```ruby
# Grab 100 appeals
appeal_ids = Task.open.where(appeal_type: LegacyAppeal.name).pluck(:appeal_id).uniq.first(100)
vacols_ids = LegacyAppeal.where(id: appeal_ids).pluck(:vacols_id)

def new_call(vacols_ids)
  VACOLS::Case.where(bfkey: vacols_ids).map do |vacols_case|
    legacy_appeal = AppealRepository.build_appeal(vacols_case) # build non-persisting legacy appeal object

    [
      vacols_case.bfkey,
      {
        location: vacols_case.bfcurloc,
        status: VACOLS::Case::TYPES[vacols_case.bfac],
        hearing_request_type: legacy_appeal.current_hearing_request_type(readable: true),
        former_travel: former_travel?(legacy_appeal)
      }
    ]
  end.to_h
end

def former_travel?(legacy_appeal)
  # the current request type is travel
  if legacy_appeal.current_hearing_request_type == :travel_board
    return false
  end

  # otherwise check if og request type was travel
  legacy_appeal.original_hearing_request_type == :travel_board
end

Benchmark.measure { new_call(vacols_ids) }
=> #<Benchmark::Tms:0x000000000e426bd0 @label="", @real=48.00420006499917, @cstime=0.0, @cutime=0.0, @stime=0.06409600000000015, @utime=1.2459829999999998, @total=1.310079>
```
1. Exit the console and reenter
```ruby
appeal_ids = Task.open.where(appeal_type: LegacyAppeal.name).pluck(:appeal_id).uniq.first(100)
vacols_ids = LegacyAppeal.where(id: appeal_ids).pluck(:vacols_id)

def new_new_call(vacols_ids)
  VACOLS::Case.where(bfkey: vacols_ids).map do |vacols_case|
    original_request = original_hearing_request_type_for_vacols_case(vacols_case)
    changed_request = changed_hearing_request_type_for_all_vacols_ids[vacols_case.bfkey]
    current_request = HearingDay::REQUEST_TYPES.key(changed_request)&.to_sym || original_request

    [
      vacols_case.bfkey,
      {
        location: vacols_case.bfcurloc,
        status: VACOLS::Case::TYPES[vacols_case.bfac],
        hearing_request_type: LegacyAppeal::READABLE_HEARING_REQUEST_TYPES[current_request],
        former_travel: original_request == :travel_board && current_request != :travel_board
      }
    ]
  end.to_h
end

def original_hearing_request_type_for_vacols_case(vacols_case)
  request_type = VACOLS::Case::HEARING_REQUEST_TYPES[vacols_case.bfhr]

  (request_type == :travel_board && vacols_case.bfdocind == "V") ? :video : request_type
end

def changed_hearing_request_type_for_all_vacols_ids
  @changed_hearing_request_type_for_all_vacols_ids ||= LegacyAppeal.pluck(:vacols_id, :changed_request_type).to_h
end

Benchmark.measure { new_new_call(vacols_ids) }
=> #<Benchmark::Tms:0x0000000021a57258 @label="", @real=12.882008823999058, @cstime=0.0, @cutime=0.0, @stime=0.2183029999999999, @utime=3.3765739999999997, @total=3.5948769999999994>
```

### CODE REVIEW

File: app/jobs/update_cached_appeals_attributes_job.rb [hunk 1]
```diff
@@ -267,20 +267,34 @@ def case_fields_for_vacols_ids(vacols_ids)
     #   ...
     # }
     VACOLS::Case.where(bfkey: vacols_ids).map do |vacols_case|
-      legacy_appeal = AppealRepository.build_appeal(vacols_case) # build non-persisting legacy appeal object
```
Comment (hschallhorn) [inline app/jobs/update_cached_appeals_attributes_job.rb:270]: This call will load the legacy appeal, vacols folder, vacols correspondent, and vacols issues for each vacols case here when the vacols case actually has almost all the information we need!

File: app/jobs/update_cached_appeals_attributes_job.rb [hunk 2]
```diff
@@ -267,20 +267,34 @@ def case_fields_for_vacols_ids(vacols_ids)
...
       ]
     end.to_h
   end
 
+  # Gets the symbolic representation of the original type of hearing requested for a vacols case record
+  def original_hearing_request_type_for_vacols_case(vacols_case)
+    request_type = VACOLS::Case::HEARING_REQUEST_TYPES[vacols_case.bfhr]
+
+    (request_type == :travel_board && vacols_case.bfdocind == "V") ? :video : request_type
+  end
```
Comment (hschallhorn) [inline app/jobs/update_cached_appeals_attributes_job.rb:291]: Replicates logic of [original_hearing_request_type](https://github.com/department-of-veterans-affairs/caseflow/blob/0324c07738bcfc62ca997e8850965bc27243f498/app/models/legacy_appeal.rb#L422)

File: app/jobs/update_cached_appeals_attributes_job.rb [hunk 3]
```diff
@@ -267,20 +267,34 @@ def case_fields_for_vacols_ids(vacols_ids)
...
+
+  # Maps vacols ids to their leagcy appeal's changed hearing request type
+  def changed_hearing_request_type_for_all_vacols_ids
+    @changed_hearing_request_type_for_all_vacols_ids = LegacyAppeal.pluck(:vacols_id, :changed_request_type).to_h
+  end
```
Comment (hschallhorn) [inline app/jobs/update_cached_appeals_attributes_job.rb:296]: single call to grab all changed_request_type for all legacy appeals

File: app/jobs/update_cached_appeals_attributes_job.rb [hunk 4]
```diff
@@ -267,20 +267,34 @@ def case_fields_for_vacols_ids(vacols_ids)
...
 
       [
         vacols_case.bfkey,
         {
           location: vacols_case.bfcurloc,
           status: VACOLS::Case::TYPES[vacols_case.bfac],
-          hearing_request_type: legacy_appeal.current_hearing_request_type(readable: true),
-          former_travel: former_travel?(legacy_appeal)
+          hearing_request_type: LegacyAppeal::READABLE_HEARING_REQUEST_TYPES[current_request],
+          former_travel: original_request == :travel_board && current_request != :travel_board
```
Comment (hschallhorn) [inline app/jobs/update_cached_appeals_attributes_job.rb:280]: replicates logic of [former_travel?](https://github.com/department-of-veterans-affairs/caseflow/pull/15478/files#diff-535f6503ed26e353aadc458806f64771b6a8ba969ede88cbb485be56f274f455L307-L316)

File: app/jobs/update_cached_appeals_attributes_job.rb [hunk 5]
```diff
@@ -267,20 +267,34 @@ def case_fields_for_vacols_ids(vacols_ids)
     #   ...
     # }
     VACOLS::Case.where(bfkey: vacols_ids).map do |vacols_case|
-      legacy_appeal = AppealRepository.build_appeal(vacols_case) # build non-persisting legacy appeal object
+      original_request = original_hearing_request_type_for_vacols_case(vacols_case)
+      changed_request = changed_hearing_request_type_for_all_vacols_ids[vacols_case.bfkey]
+      current_request = HearingDay::REQUEST_TYPES.key(changed_request)&.to_sym || original_request
```
Comment (hschallhorn) [inline app/jobs/update_cached_appeals_attributes_job.rb:272]: Replicates logic of [current_hearing_request_type](https://github.com/department-of-veterans-affairs/caseflow/blob/0324c07738bcfc62ca997e8850965bc27243f498/app/models/legacy_appeal.rb#L438)

File: app/jobs/update_cached_appeals_attributes_job.rb [hunk 6]
```diff
@@ -267,20 +267,34 @@ def case_fields_for_vacols_ids(vacols_ids)
...
         }
       ]
     end.to_h
   end
 
+  # Gets the symbolic representation of the original type of hearing requested for a vacols case record
+  def original_hearing_request_type_for_vacols_case(vacols_case)
+    request_type = VACOLS::Case::HEARING_REQUEST_TYPES[vacols_case.bfhr]
+
+    (request_type == :travel_board && vacols_case.bfdocind == "V") ? :video : request_type
```
Comment (rubaiyat22) [inline app/jobs/update_cached_appeals_attributes_job.rb:290]: very short and concise!

**Review by rubaiyat22 — APPROVED** [review]
lgtm! the logic looks correct and the tests are all passing

**Pull request merged**